// @forestadmin/forestadmin-client/dist/permissions/action-permission.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const generate_actions_from_permissions_1 = __importDefault(require("./generate-actions-from-permissions"));
class ActionPermissionService {
    constructor(options, forestAdminServerInterface) {
        this.options = options;
        this.forestAdminServerInterface = forestAdminServerInterface;
    }
    async isDevelopmentPermission() {
        const permissions = await this.getPermissions();
        // isDevelopment is true only for development environment
        return permissions.isDevelopment;
    }
    can(roleId, actionName) {
        return this.hasPermissionOrRefetch({
            roleId,
            actionName,
            // Only allow refetch when not using server events
            allowRefetch: !this.options.instantCacheRefresh,
        });
    }
    async hasPermissionOrRefetch({ roleId, actionName, allowRefetch, }) {
        const permissions = await this.getPermissions();
        const isAllowed = this.isAllowed({ permissions, actionName, roleId });
        if (!isAllowed && allowRefetch) {
            this.invalidateCache();
            return this.hasPermissionOrRefetch({
                roleId,
                actionName,
                allowRefetch: false,
            });
        }
        this.options.logger('Debug', `User ${roleId} is ${isAllowed ? '' : 'not '}allowed to perform ${actionName}`);
        return isAllowed;
    }
    isAllowed({ permissions, actionName, roleId, }) {
        // In development everything is allowed
        return Boolean(permissions.isDevelopment ||
            permissions.actionsGloballyAllowed.has(actionName) ||
            permissions.actionsByRole.get(actionName)?.allowedRoles.has(roleId));
    }
    async getPermissions() {
        if (this.permissionsPromise &&
            this.permissionExpirationTimestamp &&
            this.permissionExpirationTimestamp > Date.now()) {
            return this.permissionsPromise;
        }
        this.permissionsPromise = this.fetchEnvironmentPermissions();
        this.permissionExpirationTimestamp =
            Date.now() + this.options.permissionsCacheDurationInSeconds * 1000;
        return this.permissionsPromise;
    }
    async fetchEnvironmentPermissions() {
        this.options.logger('Debug', 'Fetching environment permissions');
        const rawPermissions = await this.forestAdminServerInterface.getEnvironmentPermissions(this.options);
        return (0, generate_actions_from_permissions_1.default)(rawPermissions);
    }
    async getCustomActionCondition(roleId, actionName) {
        const permissions = await this.getPermissions();
        const conditionFilter = permissions.actionsByRole.get(actionName)?.conditionsByRole.get(roleId);
        return conditionFilter;
    }
    async getAllCustomActionConditions(actionName) {
        const permissions = await this.getPermissions();
        return permissions.actionsByRole.get(actionName)?.conditionsByRole;
    }
    async getRoleIdsAllowedToApproveWithoutConditions(actionName) {
        const permissions = await this.getPermissions();
        const approvalPermission = permissions.actionsByRole.get(actionName);
        if (!approvalPermission) {
            return [];
        }
        // All allowed roles excluding the one with conditions
        return Array.from(approvalPermission.allowedRoles).filter(roleId => !approvalPermission.conditionsByRole?.has(roleId));
    }
    invalidateCache() {
        this.options.logger('Debug', 'Invalidating roles permissions cache..');
        this.permissionsPromise = undefined;
        this.permissionExpirationTimestamp = undefined;
    }
}
exports.default = ActionPermissionService;
