// @forestadmin/forestadmin-client/dist/permissions/rendering-permission.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lru_cache_1 = __importDefault(require("lru-cache"));
const hash_chart_1 = require("./hash-chart");
const is_segment_query_authorized_1 = __importDefault(require("./is-segment-query-authorized"));
const types_1 = require("./types");
const verify_sql_query_1 = __importDefault(require("./verify-sql-query"));
const context_variables_1 = __importDefault(require("../utils/context-variables"));
const context_variables_injector_1 = __importDefault(require("../utils/context-variables-injector"));
class RenderingPermissionService {
    constructor(options, userPermissions, forestAdminServerInterface) {
        this.options = options;
        this.userPermissions = userPermissions;
        this.forestAdminServerInterface = forestAdminServerInterface;
        this.permissionsByRendering = new lru_cache_1.default({
            max: 256,
            ttl: this.options.permissionsCacheDurationInSeconds * 1000,
            fetchMethod: renderingId => this.loadPermissions(Number(renderingId)),
        });
    }
    async getScope({ renderingId, collectionName, userId, }) {
        return this.getScopeOrRetry({
            renderingId,
            collectionName,
            userId,
            // Only allow retry when not using server events
            allowRetry: !this.options.instantCacheRefresh,
        });
    }
    async getScopeOrRetry({ renderingId, collectionName, userId, allowRetry, }) {
        const [permissions, userInfo] = await Promise.all([
            this.permissionsByRendering.fetch(`${renderingId}`),
            this.userPermissions.getUserInfo(userId),
        ]);
        const collectionPermissions = permissions?.collections?.[collectionName];
        if (!collectionPermissions) {
            if (allowRetry) {
                this.invalidateCache(renderingId);
                return this.getScopeOrRetry({
                    renderingId,
                    collectionName,
                    userId,
                    allowRetry: false,
                });
            }
            return null;
        }
        return context_variables_injector_1.default.injectContextInFilter(collectionPermissions.scope, new context_variables_1.default({ team: permissions.team, user: userInfo }));
    }
    async canExecuteSegmentQuery({ renderingId, collectionName, segmentQuery, }) {
        return ((await this.canExecuteSegmentQueryOrRetry({
            renderingId,
            collectionName,
            segmentQuery,
            // Only allow retry when not using server events
            allowRetry: !this.options.instantCacheRefresh,
        })) && (0, verify_sql_query_1.default)(segmentQuery));
    }
    async canExecuteSegmentQueryOrRetry({ renderingId, collectionName, segmentQuery, allowRetry, }) {
        const permissions = await this.permissionsByRendering.fetch(`${renderingId}`);
        const collectionPermissions = permissions?.collections?.[collectionName];
        if (!collectionPermissions ||
            !(0, is_segment_query_authorized_1.default)(segmentQuery, collectionPermissions.segments)) {
            if (allowRetry) {
                this.invalidateCache(renderingId);
                return this.canExecuteSegmentQueryOrRetry({
                    renderingId,
                    collectionName,
                    segmentQuery,
                    allowRetry: false,
                });
            }
            this.options.logger('Debug', `User cannot retrieve SQL segment on rendering ${renderingId}`);
            return false;
        }
        this.options.logger('Debug', `User can retrieve SQL segment on rendering ${renderingId}`);
        return true;
    }
    async loadPermissions(renderingId) {
        this.options.logger('Debug', `Loading rendering permissions for rendering ${renderingId}`);
        const rawPermissions = await this.forestAdminServerInterface.getRenderingPermissions(renderingId, this.options);
        const charts = (0, hash_chart_1.hashServerCharts)(rawPermissions.stats);
        return {
            team: rawPermissions.team,
            collections: rawPermissions.collections,
            charts,
        };
    }
    isQueryChart(chartRequest) {
        return 'query' in chartRequest;
    }
    async canExecuteChart({ renderingId, chartRequest, userId, }) {
        const chartHash = (0, hash_chart_1.hashChartRequest)(chartRequest);
        return ((await this.canRetrieveChartHashOrRetry({
            renderingId,
            chartHash,
            userId,
            // Only allow retry when not using server events
            allowRetry: !this.options.instantCacheRefresh,
        })) &&
            (!this.isQueryChart(chartRequest) || (0, verify_sql_query_1.default)(chartRequest.query)));
    }
    async canRetrieveChartHashOrRetry({ renderingId, userId, chartHash, allowRetry, }) {
        const [userInfo, permissions] = await Promise.all([
            this.userPermissions.getUserInfo(userId),
            this.permissionsByRendering.fetch(`${renderingId}`),
        ]);
        if ([types_1.PermissionLevel.Admin, types_1.PermissionLevel.Developer, types_1.PermissionLevel.Editor].includes(userInfo?.permissionLevel) ||
            permissions.charts.has(chartHash)) {
            this.options.logger('Debug', `User ${userId} can retrieve chart on rendering ${renderingId}`);
            return true;
        }
        if (allowRetry) {
            this.invalidateCache(renderingId);
            this.userPermissions.invalidateCache();
            return this.canRetrieveChartHashOrRetry({
                renderingId,
                userId,
                chartHash,
                allowRetry: false,
            });
        }
        this.options.logger('Debug', `User ${userId} cannot retrieve chart on rendering ${renderingId}`);
        return false;
    }
    invalidateCache(renderingId) {
        this.options.logger('Debug', `Invalidating rendering permissions cache for rendering ${renderingId}`);
        this.permissionsByRendering.delete(`${renderingId}`);
    }
    invalidateAllCache() {
        this.options.logger('Debug', `Invalidating rendering permissions cache for all renderings`);
        this.permissionsByRendering.clear();
    }
    async getUser(userId) {
        return this.userPermissions.getUserInfo(userId);
    }
    async getTeam(renderingId) {
        const permissions = await this.permissionsByRendering.fetch(`${renderingId}`);
        return permissions.team;
    }
}
exports.default = RenderingPermissionService;
