// @forestadmin/forestadmin-client/dist/forest-admin-client-with-cache.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const verify_approval_1 = __importDefault(require("./permissions/verify-approval"));
class ForestAdminClientWithCache {
    constructor(options, permissionService, renderingPermissionService, contextVariablesInstantiator, chartHandler, ipWhitelistService, schemaService, authService, eventsSubscription) {
        this.options = options;
        this.permissionService = permissionService;
        this.renderingPermissionService = renderingPermissionService;
        this.contextVariablesInstantiator = contextVariablesInstantiator;
        this.chartHandler = chartHandler;
        this.ipWhitelistService = ipWhitelistService;
        this.schemaService = schemaService;
        this.authService = authService;
        this.eventsSubscription = eventsSubscription;
    }
    verifySignedActionParameters(signedParameters) {
        return (0, verify_approval_1.default)(signedParameters, this.options.envSecret);
    }
    getIpWhitelistConfiguration() {
        return this.ipWhitelistService.getConfiguration();
    }
    async postSchema(schema) {
        return this.schemaService.postSchema(schema);
    }
    getOpenIdClient() {
        return this.authService.getOpenIdClient();
    }
    getUserInfo(renderingId, accessToken) {
        return this.authService.getUserInfo(renderingId, accessToken);
    }
    async getScope({ renderingId, userId, collectionName, }) {
        return this.renderingPermissionService.getScope({
            renderingId,
            collectionName,
            userId,
        });
    }
    markScopesAsUpdated(renderingId) {
        if (!this.options.instantCacheRefresh)
            this.renderingPermissionService.invalidateCache(renderingId);
    }
    async subscribeToServerEvents() {
        await this.eventsSubscription.subscribeEvents();
    }
}
exports.default = ForestAdminClientWithCache;
