// @forestadmin/forestadmin-client/dist/events-subscription/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const eventsource_1 = __importDefault(require("eventsource"));
const types_1 = require("./types");
class EventsSubscriptionService {
    constructor(options, refreshEventsHandlerService) {
        this.options = options;
        this.refreshEventsHandlerService = refreshEventsHandlerService;
    }
    async subscribeEvents() {
        if (!this.options.instantCacheRefresh) {
            this.options.logger('Debug', 'Event source deactivated.. Use agent option [instantCacheRefresh=true] ' +
                'if you want to activate them');
            return;
        }
        const eventSourceConfig = {
            // forest-secret-key act as the credential
            withCredentials: false,
            headers: { 'forest-secret-key': this.options.envSecret },
            https: {
                rejectUnauthorized: process.env.NODE_TLS_REJECT_UNAUTHORIZED !== '0',
            },
        };
        const url = new URL('/liana/v4/subscribe-to-events', this.options.forestServerUrl).toString();
        const source = new eventsource_1.default(url, eventSourceConfig);
        source.addEventListener('error', this.onEventError.bind(this));
        source.addEventListener('open', () => this.onEventOpen());
        source.addEventListener(types_1.ServerEventType.RefreshUsers, async () => this.refreshEventsHandlerService.refreshUsers());
        source.addEventListener(types_1.ServerEventType.RefreshRoles, async () => this.refreshEventsHandlerService.refreshRoles());
        source.addEventListener(types_1.ServerEventType.RefreshRenderings, async (event) => this.handleSeverEventRefreshRenderings(event));
        source.addEventListener(types_1.ServerEventType.RefreshCustomizations, async () => this.refreshEventsHandlerService.refreshCustomizations());
    }
    async handleSeverEventRefreshRenderings(event) {
        if (!event.data) {
            this.options.logger('Debug', 'Server Event - RefreshRenderings missing required data.');
            return;
        }
        const { renderingIds } = JSON.parse(event.data);
        await this.refreshEventsHandlerService.refreshRenderings(renderingIds);
    }
    onEventError(event) {
        if (event.status === 502) {
            this.options.logger('Debug', 'Server Event - Connection lost');
            return;
        }
        if (event.message)
            this.options.logger('Warn', `Server Event - Error: ${JSON.stringify(event)}`);
    }
    onEventOpen() {
        this.options.logger('Debug', 'Server Event - Open EventSource (SSE) connection with Forest Admin servers');
        // Flush all previous data as we could have missed some events
        this.refreshEventsHandlerService.refreshEverything();
    }
}
exports.default = EventsSubscriptionService;
